packages/adapters-library/src/adapters/metamask/products/pooled-staking/metamaskPooledStakingAdapter.ts

```diff
@@ -0,0 +1,167 @@
+
+const METAMASK_POOLED_STAKING_CONTRACT = getAddress(
+  '0x4FEF9D741011476750A243aC70b9789a63dd47Df',
+)
+const LIDO_STAKED_ETH = getAddress('0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84')
+
+export class MetamaskPooledStakingAdapter implements IProtocolAdapter {
+  productId = 'pooled-staking'
+  protocolId: Protocol
+  chainId: Chain
+  helpers: Helpers
+
+  adapterSettings: AdapterSettings = {
+    includeInUnwrap: true,
+    userEvent: {
+      topic0:
+        '0x861a4138e41fb21c121a7dbb1053df465c837fc77380cc7226189a662281be2c',
+      userAddressIndex: 1,
+    },
+  }
+
+  private provider: CustomJsonRpcProvider
+
+  adaptersController: AdaptersController
+
+  private poolContract: PoolStaking
+
+  constructor({
+    provider,
+    chainId,
+    protocolId,
+    adaptersController,
+    helpers,
+  }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+    this.helpers = helpers
+
+    this.poolContract = PoolStaking__factory.connect(
+      METAMASK_POOLED_STAKING_CONTRACT,
+      this.provider,
+    )
+  }
+
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'Metamask',
+      description: 'Metamask defi adapter',
+      siteUrl: 'https://portfolio.metamask.io/',
+      iconUrl: 'https://portfolio.metamask.io/favicon.png',
+      positionType: PositionType.Supply,
+      chainId: this.chainId,
+      productId: this.productId,
+    }
+  }
+
+  @CacheToDb
+  async getProtocolTokens(): Promise<ProtocolToken[]> {
+    const underlyingToken = await this.helpers.getTokenMetadata(LIDO_STAKED_ETH)
+    return [
+      {
+        address: METAMASK_POOLED_STAKING_CONTRACT,
+        symbol: 'METAMASK',
+        name: 'Metamask',
+        decimals: 18,
+        underlyingTokens: [underlyingToken],
+      },
+    ]
+  }
+
+  async getPositions(input: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const protocolTokens = await this.getProtocolTokens()
+
+    if (protocolTokens.length !== 1) {
+      throw new Error('Only one protocol token is supported')
+    }
+
+    const { underlyingTokens, ...protocolToken } = protocolTokens[0]!
+
+    const balance = await this.poolContract.getShares(input.userAddress)
+
+    return [
+      {
+        ...protocolToken,
+        type: TokenType.Protocol,
+        balanceRaw: balance,
+      },
+    ]
+  }
+
+  async unwrap({
+    protocolTokenAddress,
+    tokenId,
+    blockNumber,
+  }: UnwrapInput): Promise<UnwrapExchangeRate> {
+    const protocolTokens = await this.getProtocolTokens()
+
+    if (protocolTokens.length !== 1) {
+      throw new Error('Only one protocol token is supported')
+    }
+
+    const { underlyingTokens, ...protocolToken } = protocolTokens[0]!
+
+    if (underlyingTokens.length !== 1) {
+      throw new Error('Only one underlying token is supported')
+    }
```

Comment:
I'm not a big fan of this unreachable throws. If we know what the shape of the adapter metadata is, we can just go with `!`.

Alternative, doing an `assert(condition, errorMessage)` would be cleaner and won't decrease code coverage.